test/e2e/certificates_test.go

```diff
@@ -0,0 +1,366 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"path/filepath"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/openshift/cert-manager-operator/test/library"
+	infrav1 "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+
+	v1 "github.com/cert-manager/cert-manager/pkg/apis/acme/v1"
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	certmanagermetav1 "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	certmanagerclientset "github.com/cert-manager/cert-manager/pkg/client/clientset/versioned"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("ACME Certificate", Ordered, func() {
+	t := &testing.T{}
+
+	var ctx context.Context
+	var loader library.DynamicResourceLoader
+
+	var routeClient *library.RoutesClient
+	var infraClient *infrav1.ConfigV1Client
+	var certmanagerClient *certmanagerclientset.Clientset
+
+	var appsDomain string
+
+	BeforeAll(func() {
+		By("creating Kube clients")
+		ctx = context.Background()
+		loader = library.NewDynamicResourceLoader(ctx, t)
+
+		config, err := library.GetConfigForTest(t)
+		Expect(err).NotTo(HaveOccurred())
+
+		routeClient = library.NewRoutesClient(t, ctx)
+		appsDomain = routeClient.GetIngressAppsDomain()
+		Expect(appsDomain).NotTo(BeEmpty())
+
+		infraClient, err = infrav1.NewForConfig(config)
+		Expect(err).NotTo(HaveOccurred())
+
+		certmanagerClient, err = certmanagerclientset.NewForConfig(config)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	BeforeEach(func() {
+		By("waiting for operator status to become available")
+		err := waitForValidOperatorStatusCondition(certmanageroperatorclient,
+			[]string{certManagerControllerDeploymentControllerName,
+				certManagerWebhookDeploymentControllerName,
+				certManagerCAInjectorDeploymentControllerName})
+		Expect(err).NotTo(HaveOccurred(), "Operator is expected to be available")
+	})
+
+	Context("dns-01 challenge using explicit credentials", func() {
+		It("should obtain a valid LetsEncrypt certificate", func() {
+
+			By("creating a test namespace")
+			ns, err := loader.CreateTestingNS("e2e-acme-explicit-dns01")
+			Expect(err).NotTo(HaveOccurred())
+			defer loader.DeleteTestingNS(ns.Name)
```

Comment:
Checking error(s) for the defer call, makes the code look like the following:
```
// could be placed in utils.go
func AckCondition(condFunc func() (bool, error), errHandleFunc func(error)) {
	cond, err := condFunc()
	if !cond {
		errHandleFunc(err)
	}
}
```
with the actual defer call as:
```
defer library.AckCondition(func() (bool, error) {
	return loader.DeleteTestingNS(ns.Name)
}, func(err error) {
	Expect(err).NotTo(HaveOccurred())
})
```

Also, there is slight pain in directly checking the error with the Delete* func(s) from library/utils, because their implementation returns the error of a Get call performed after actual delete of the resource. Irrespective of it's a k8s not found, the error is returned. Currently, that feels bit counter intuitive as it should only return an error when deletion have failed but am not willing to change the whole implementation since it has been there for quite sometime and is well tested in practice.

Let me know if you have an alternate error checking strategy with this or if we could skip the error test in the all the defer delete calls altogether (like the earlier tests did).

Reply:
cc: @thejasn

Reply:
If we are using defers to **only** delete NS, we skip the error check, seems a bit extra IMO.